#22 OmSupply windows installers

raviSussol opened the pull request:
Closes #22

clemens-msupply reviewed (commented):
Thanks! only did a quick scan through it since I have no idea about how the actual installer works. Just one comment, maybe rename the `build` directory to `installer_win`?

raviSussol commented on server/server/src/set_icon.rs:
```diff
@@ -0,0 +1,15 @@
+fn main() {
+    let mut res = winres::WindowsResource::new();
+    res.set_icon("omSupply.ico");
+    res.compile().unwrap();
```
cc30919a3

raviSussol commented on server/server/src/set_icon.rs:
```diff
@@ -0,0 +1,15 @@
+
+// Just to skip build error as the build dependencies for "winres" is being used for "windows" targeted
+// This way the build time for the unix based platform gets saved
+#[cfg(unix)]
```
Right! Just to make unix build happy. cc30919a3

raviSussol commented:
> Thanks! only did a quick scan through it since I have no idea about how the actual installer works. Just one comment, maybe rename the `build` directory to `installer_win`?

Hi @clemens-msupply Updated the `build` folder to `installer_win`. Basically when the new tag is pushed by someone to the Github then the Jenkins (which I've configured for the open-msupply in our Jenkins server in the cloud) would trigger the [omSupplyMain](https://jenkins.msupply.org:8443/job/omSupplyMain/) job for compilation, building of the tagged project. After the `omSupplyMain` job finishes it trigger the another `omSupplyMain - Installers` job which is basically responsible to create windows installers using the Setup factory based on the templates provided (the files with `*.suf` in the `installer_win` folder) in the Jenkins server. You might want to check out as an example built windows installers in [here](https://jenkins.msupply.org:8443/job/omSupplyMain%20-%20Installers/)

raviSussol commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Installers openmsupply](https://trello.com/c/fLP85QlU/549-installers-openmsupply)

mark-prins commented on installer_win/omsupply_web_server.suf:
```diff
@@ -0,0 +1,4455 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
:nit the name of the suf file isn't quite accurate: the remote server is more than a web server, so the suf files would be better named just `omsupply_server.suf` and `omsupply_server_upgrader.suf`

raviSussol replied:
I was trying to emphasize that the installer is for web interface based server so used "web" word in it - though it does sounds like or confuses that it is just only a normal web server.
5c5269893

mark-prins commented on installer_win/omsupply-desktop-build.bat:
```diff
@@ -0,0 +1,3 @@
+@ECHO ##### Building omSupply for the desktop #####
+SET installerWorkspace=C:\Program Files (x86)\Jenkins\jobs\omSupplyMain - installers\workspace\omSupply
+cd .. && cd "server" && cargo clean && cd .. && cd "client" && yarn install --force && yarn electron:build && cd .. && @ECHO ##### Copying build artifacts to the installer folder ##### && xcopy "client\packages\electron\out\open mSupply-win32-x64\**" "%installerWorkspace%\omSupply Desktop Server\" /e /h /c /i
```
there's no need to run cargo clean when building the electron app

raviSussol replied:
755560e30

mark-prins commented on installer_win/omsupply-postgres-build.bat:
```diff
@@ -0,0 +1,3 @@
+@ECHO ##### Building omsupply for the postgres #####
+SET installerWorkspace=C:\Program Files (x86)\Jenkins\jobs\omSupplyMain - installers\workspace\omSupply
+cd .. && cd "server" && cargo clean && cd .. && cd "client" && yarn install --force && yarn build && cd .. && cd "server" && cargo build --release --features postgres && cd .. && @ECHO ##### Copying build artifacts to the installer folder ##### && copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Web Server\omSupply-postgres.exe" && copy "server\target\release\remote_s ...
```
the command would be slightly neater if you just went into the client dir first, built the client then went ../server and ran all the server commands.

e.g.
```suggestion
cd .. && cd client && yarn install --force && yarn build && cd .. && cd server && cargo clean && cargo build --release --features postgres && cd .. && @ECHO ##### Copying build artifacts to the installer folder ##### && copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Web Server\omSupply-postgres.exe" && copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Desktop Server\server\omSupply-postgres.exe"
```

mark-prins reviewed (commented):
Looking good thanks! 
Just had a read through of what I can - will need a windows machine to test the installers and look at the SF scripts more carefully.

mark-prins commented on server/configuration/base.yaml:
```diff
@@ -1,5 +1,6 @@
+# Replace [[xxx]] with valid values in their corresponding fields. See side comments for an example below.
 server:
-  port: 8000
+  port: [[omsupply_port]] # e.g. 8000
```
I think the defaults are helpful - for devs it means that you just clone the repo and you should be able to run the project. I'm guessing that you're doing a replace with setup factory, which is why the change? If so - is is possible to use regex and search for `port:[\w\W]*/n` for example?

raviSussol replied:
Yes to make easy for setup factory to replace old config details in yaml file. Updated with 82988f868

mark-prins commented on installer_win/omsupply-build.bat:
```diff
@@ -0,0 +1,26 @@
+copy "version.txt" "%installerWorkspace%\version.txt"
+
+@cd "installer_win"
+start /b /wait omsupply-sqlite-build.bat
```
I was wondering if we could speed up the build - one aspect is to remove the duplication, there is multiple `yarn install` and `yarn build` and `cargo clean` which would not be needed if you didn't have the builds split out into separate batch files.

I would also run separate lines and take the risk that, for example, the yarn build fails and the cargo build doesn't embed the client - though actually the cargo build would fail in that case..

Also not sure why `cd installer_win` and then `cd ..` is needed?

```
SET installerWorkspace=C:\Program Files (x86)\Jenkins\jobs\omSupplyMain - installers\workspace\omSupply
cd client && yarn install --force && yarn build
cd server && cargo clean

@ECHO ##### Building omSupply for the desktop #####
cd client && yarn electron:build

@ECHO ##### Copying build artifacts to the installer folder ##### 
xcopy "client\packages\electron\out\open mSupply-win32-x64\**" "%installerWorkspace%\omSupply Desktop Server\" /e /h /c /i

@ECHO ##### Building omsupply for the postgres #####
cd server && cargo build --release --features postgres

@ECHO ##### Copying build artifacts to the installer folder ##### 
copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Web Server\omSupply-postgres.exe" && copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Desktop Server\server\omSupply-postgres.exe"

@ECHO ##### Building omsupply for the sqlite #####
cd "server" && cargo build --release 

@ECHO ##### Copying build artifacts to the installer folder ##### 
copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Web Server\omSupply-sqlite.exe" && copy "server\target\release\remote_server.exe" "%installerWorkspace%\omSupply Desktop Server\server\omSupply-sqlite.exe"
```

raviSussol replied:
Yeah it would have been clean and simple if the build scrips could be written in a single bat file and I had tried that as well but the bat file execution didn't seem to be run in sequential manner while I had been testing on windows. If the command which likely to take long time for execution then the bat execution simply doesn't wait for it to be finished instead the execution moves to the next command line which makes the build failed for an empty omsupply exe files for setup factory to make installers.

So had to split each individual build into a separate bat file with `wait` flag to start their execution. I also had to chain the series of commands in each separate bat files so that the command executions don't get broken when first command line executes.

mark-prins replied:
oh, that's disappointing. I had a similar problem on jenkins, now that you mention it.
Currently using a bat file on the demo server to build.. but that's windows command line, not jenkins 🙄

mark-prins reviewed (commented):
and, just quietly, I preferred the `build` directory name. I would have gone with `build/windows` with the shared items ( only the license, I think? ) in build and all of the windows specific bits in the windows directory.. but 🤷  either is fine!

mark-prins reviewed (approved):
🎉  Looks great - thanks!

I've added a small change in 3f73c3b4a because I thought I could speed up the main build a bit and wanted to test with jenkins. Had some issues running the installers, which I think is because the suf files are run from the lower level directory, so changed that a bit too.

I've also removed `nssm` usage - as Craig noted in the issue, we can use setup factory for that ( now that we have logging in the server!)

andreievg commented on build/windows/omsupply-prepare.bat:
```diff
@@ -0,0 +1,3 @@
+@ECHO ##### Prepare omsupply build #####
+REM cd "..\..\client" && yarn install --force && yarn build && cd "..\server" && cargo clean
+cd "..\..\client" && yarn install --force && yarn build && cd "..\server" && cargo clean
```
i don't think we need to  do `cargo clean` here, and also `yarn install --force --frozen-lockfile`

mark-prins replied:
thanks Andrei - popped that in too 😉

raviSussol commented:
@mark-prins 
> I've added a small change in https://github.com/openmsupply/open-msupply/commit/3f73c3b4a7236a640a6ae17426898a40c191e00d because I thought I could speed up the main build a bit and wanted to test with jenkins.

👍 

> Had some issues running the installers, which I think is because the suf files are run from the lower level directory, so changed that a bit too.

That's strange... I'd tested them all in 2 windows pcs in my last successful builds installers and they all were working fine.

> I've also removed nssm usage - as Craig noted in the issue, we can use setup factory for that ( now that we have logging in the server!)

I've already tested creating service with setup factory script and tried to run it - doesn't work at all. To make working we have to handle service management within our remote server app by ourselves. It is the same as the cmd command `sc create <service_name> binpath= "service_full_path"` which expects the app you're trying to run it as a service must have handled service management internally in the app by software vendor at first place before you create for it to run as service.

Hence I had to rely back on the `nssm` for service management.

mark-prins commented:
> Hence I had to rely back on the `nssm` for service management.

ah, right. I see now - that's a pity. we might have to look later and see if we can handle the service methods in rust. I've only done that in C#

raviSussol commented:
> ah, right. I see now - that's a pity. we might have to look later and see if we can handle the service methods in rust. I've only done that in C#

Yeah I had done a quick research for service management in rust and had found a crate that handle service management on windows. But seemed to me that it needs a bit of researching and a bit amount of testing as well - so didn't bother to work on that for now. I'll revert your particular change in suf files. Plus I saw you have removed `app_data` folder link in suf files thinking that it would be created when the app gets run. But the reason that I'd added `app_data\certs` folder in suf files for comprehensive for users or developers like us so that when we need to put cert files after the installation it would becomes easy for us to navigate where we need to simply put them. So is it OK if I revert that too?

mark-prins commented:
yeah, thanks @raviSussol - the `app_data\certs` was mainly because we shouldn't have certificates in the repo / bundled with the installer. Best just to revert those bits.. I'll keep my fingers out of things! I think that you've got it covered and I'm just working through things that you've already thought of.

Agree that the service management crate needs a bit of research and testing - it looks promising and overall that would be neater than using nssm, so I think it's worth working on. Just not right now, as you say.

raviSussol commented:
All good now and good to be merged.
